controllers/networkchaos/types.go

```diff
@@ -41,16 +48,54 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, nil
 	}
 
+	scheduler := networkchaos.GetScheduler()
+	duration, err := networkchaos.GetDuration()
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to get podchaos[%s/%s]'s duration", networkchaos.Namespace, networkchaos.Name))
+		return ctrl.Result{}, nil
+	}
+	if scheduler == nil && duration == nil {
+		return r.commonNetworkChaos(&networkchaos, req)
+	} else if scheduler != nil && duration != nil {
+		return r.scheduleNetworkChaos(&networkchaos, req)
+	}
+
+	// This should be ensured by admission webhook in the future
+	r.Log.Error(fmt.Errorf("networkchaos[%s/%s] spec invalid", networkchaos.Namespace, networkchaos.Name), "scheduler and duration should be omitted or defined at the same time")
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) commonNetworkChaos(networkchaos *v1alpha1.NetworkChaos, req ctrl.Request) (ctrl.Result, error) {
 	switch networkchaos.Spec.Action {
 	case v1alpha1.DelayAction, v1alpha1.DuplicateAction, v1alpha1.CorruptAction, v1alpha1.LossAction:
-		reconciler := netem.NewReconciler(r.Client, r.Log.WithValues("action", "netem"), req)
+		r := commonNetem.NewCommonReconciler(r.Client, r.Log.WithValues("action", "netem"), req)
+		reconciler := common.NewReconciler(r, r.Client, r.Log)
 		return reconciler.Reconcile(req)
 	case v1alpha1.PartitionAction:
-		reconciler := partition.NewReconciler(r.Client, r.Log.WithValues("action", "partition"), req)
+		r := commonPartition.NewCommonReconciler(r.Client, r.Log.WithValues("action", "partition"), req)
+		reconciler := common.NewReconciler(r, r.Client, r.Log)
 		return reconciler.Reconcile(req)
 	default:
-		r.Log.Error(nil, "networkchaos action is invalid", "action", networkchaos.Spec.Action)
+		return r.defaultResponse(networkchaos), nil
+	}
+}
 
-		return ctrl.Result{}, nil
+func (r *Reconciler) scheduleNetworkChaos(networkchaos *v1alpha1.NetworkChaos, req ctrl.Request) (ctrl.Result, error) {
+	switch networkchaos.Spec.Action {
+	case v1alpha1.DelayAction, v1alpha1.DuplicateAction, v1alpha1.CorruptAction, v1alpha1.LossAction:
+		r := twophaseNetem.NewTwoPhaseReconciler(r.Client, r.Log.WithValues("action", "netem"), req)
+		reconciler := twophase.NewReconciler(r, r.Client, r.Log)
```

Comment:
nit: if you want to save two lines you can

```go
var inner twohpase.InnerReconciler
swtich {
case Delay
     // assume previous comment is applied and twophaseNetem package is gone
     inner = netem.NewTwoPhaseReconciler
case Partition:
     inner = partition.NewTwoPhaseReconciler
default:
     return r.ignoreInvalidAction
}
rc := twophase.NewReconciler(inner, ...)
return rc.Reconcile
```

Reply:
updated.